File: Python/853.py
# Notes:
    # Use ETA's instead of computing relative velocities to avoid zero division
    # Another option is to just push the ETA to the stack instead of p/s
    # For each car, compute how long it will take to reach the target
        # If shorter than the top of the stack, it becomes part of a fleet
        # If longer than the top of the stack, it stays behind
    # Return the length of the stack to find the number of fleets
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        
        fleet = sorted(zip(position, speed), key = lambda x: -x[0])
        stack = []
        for (p, s) in fleet:
            ETA = (target - p) / s
            if not stack: stack.append(ETA)
            if ETA > stack[-1]: stack.append(ETA)
        
        return len(stack)

File: C++/853.cpp
/*
- Can make this a bit easier by constructing ETA's in initial array
*/
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        vector<pair<int, int>> arr;
        for(int i = 0; i < position.size(); i++){
            arr.push_back({position[i], speed[i]});
        }
        sort(arr.begin(), arr.end());
        reverse(arr.begin(), arr.end());

        stack<double> stk;
        for(int i = 0; i < arr.size(); i++){
            double ETA = (double) (target - arr[i].first) / arr[i].second;
            if(stk.empty() || ETA > stk.top()) stk.push(ETA);
        }
        
        return stk.size();
    }
};